Sources/Basic/FileSystem.swift

```diff
@@ -276,6 +239,57 @@ private class LocalFileSystem: FileSystem {
             break
         }
     }
+
+    /// An iterator to move though contents of a directory.
+    private final class DirectoryContentsIterator: Sequence, IteratorProtocol {
+        private let dir: DirHandle?
+        private let path: AbsolutePath
+        private var entry = dirent()
+
+        private init(at path: AbsolutePath) throws {
+            guard let dir = libc.opendir(path.asString) else {
+                throw FileSystemError(errno: errno)
+            }
+            self.dir = dir
+            self.path = path
+        }
+
+        deinit { _ = libc.closedir(dir) }
+
+        public func next() -> String? {
+            while true {
+                var entryPtr: UnsafeMutablePointer<dirent>? = nil
+                if readdir_r(dir, &entry, &entryPtr) < 0 {
+                    // FIXME: Are there ever situation where we would want to
+                    // continue here?
+                    // FIXME: We're ignoring the error here, find a way to notify clients.
```

Comment:
The downside of that is that it doesn't conform to `Sequence` anymore, so `map` and all that go away. It may be this iterator is special enough we don't really care...